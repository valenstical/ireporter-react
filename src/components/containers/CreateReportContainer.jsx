import React, { Component } from 'react';
import { connect } from 'react-redux';
import formDataJSON from 'formdata-json';
import PropTypes from 'prop-types';
import { createStructuredSelector } from 'reselect';
import { Redirect } from 'react-router-dom';
import { getUser, getReport } from '../../selectors';
import { scrollTop } from '../../utils/helpers';
import CreateReport from '../views/CreateReport';
import reportIncident from '../../actions/reportAction';


/**
 * Container component for the create report view
 * @export
 * @class CreateReport
 * @extends {Component}
 */
class CreateReportContainer extends Component {
  async componentDidMount() {
    const { location: { pathname } } = this.props;
    const pathAsArray = pathname.split('/');
    const reportType = pathAsArray[2];
    const reportId = pathAsArray[3];
    const { reportIncident: processReportIncident } = this.props;
    if (reportType && reportId) await processReportIncident({ route: `${reportType}/${reportId}`, method: 'get' });
    const { report } = this.props;
    const locatonAsArray = report.location.split(',');
    const [latitude, longitude] = locatonAsArray;
    report.latitude = latitude;
    report.longitude = longitude;
  }

  handleSubmit = async (event) => {
    event.preventDefault();
    const payload = formDataJSON(new FormData(event.target));
    payload.route = `${payload.type}s`;
    payload.method = payload.id ? 'patch' : 'post';
    payload.location = `${payload.latitude},${payload.longitude}`;
    const { reportIncident: processReportIncident } = this.props;
    await processReportIncident(payload);
    scrollTop();
  }

  render() {
    const { user, report } = this.props;
    if (!user.isLoggedIn) return <Redirect to="/login" />;
    if (report.success) return <Redirect to="/dashboard" />;
    if (report.redirect) return <Redirect to="/404" />;

    return (
      <CreateReport
      handleSubmit={this.handleSubmit}
      report={report}
      />
    );
  }
}
CreateReportContainer.propTypes = {
  user: PropTypes.object.isRequired,
  report: PropTypes.object.isRequired,
  location: PropTypes.object.isRequired,
  reportIncident: PropTypes.func.isRequired,
};

const mapStateToProps = createStructuredSelector(
  {
    user: getUser,
    report: getReport
  }
);

const mapDispatchToProps = dispatch => ({
  reportIncident: payload => dispatch(reportIncident(payload)),
});

export default connect(mapStateToProps, mapDispatchToProps)(CreateReportContainer);
