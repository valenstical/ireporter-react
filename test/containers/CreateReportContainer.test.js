import React from 'react';
import setup from '../setup';
import CreateReportContainer from '../../src/components/containers/CreateReportContainer';


describe('<CreateReportContainer>', () => {
  it('should render without crashing', () => {
    const wrapper = setup(<CreateReportContainer />);
    expect(wrapper).toMatchSnapshot();
  });
});
