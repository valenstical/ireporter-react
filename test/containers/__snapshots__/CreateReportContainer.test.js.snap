// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateReportContainer> should render without crashing 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter
    keyLength={0}
  >
    <Router
      history={
        Object {
          "action": "REPLACE",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/login",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Connect(CreateReportContainer)>
        <CreateReportContainer
          report={
            Object {
              "isBusy": false,
              "message": Array [],
              "success": false,
            }
          }
          reportIncident={[Function]}
          user={
            Object {
              "isBusy": false,
              "message": Array [],
              "success": false,
              "token": "",
            }
          }
        >
          <Redirect
            push={false}
            to="/login"
          />
        </CreateReportContainer>
      </Connect(CreateReportContainer)>
    </Router>
  </BrowserRouter>
</Provider>
`;
